### PR TITLE
Refactor the `MirPass for QualifyAndPromoteConstants`

### PR DESCRIPTION
**Pull request by Centril**
This is an accumulation of drive-by commits while working on `Vec::new` as a stable `const fn`.
The PR is probably easiest read commit-by-commit.

r? @oli-obk 

cc @eddyb @ecstatic-morse -- your two PRs https://github.com/rust-lang/rust/pull/63812 and https://github.com/rust-lang/rust/pull/63860 respectively will conflict with this a tiny bit but it should be trivial to reintegrate your changes atop of this.

### CODE REVIEW

**Comment by oli-obk** [comment]
This PR will conflict heavily with the mentioned two PRs. I'd like to block it on getting the others merged

**Comment by Centril** [comment]
@oli-obk I checked the diff and it will only conflict with each of those PRs in 3-7 lines at most

**Comment by Centril** [comment]
Rebased & simplified things a bit.

**Comment by JohnCSimon** [comment]
Ping from triage: 
@oli-obk @varkor @spastorino 
Hi! 
Can you please review this PR? 
Thanks

**Review by spastorino — APPROVED** [review]

File: src/librustc_mir/transform/qualify_consts.rs [hunk 1]
```diff
@@ -1664,107 +1639,119 @@ impl<'tcx> MirPass<'tcx> for QualifyAndPromoteConstants<'tcx> {
...
+    }
+}
+
+/// In `const` and `static` everything without `StorageDead`
+/// is `'static`, we don't have to create promoted MIR fragments,
+/// just remove `Drop` and `StorageDead` on "promoted" locals.
+fn remove_drop_and_storage_dead_on_promoted_locals(
+    body: &mut Body<'tcx>,
+    promoted_temps: &BitSet<Local>,
+) {
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Btw, according to @RalfJung and @oli-obk's discussions elsewhere(?), we shouldn't even be doing this (but instead run regular promotion).

**Comment by eddyb** [comment]
Would be nice if the PR author could toggle the default "whitespace" diff setting (the diff is way more readable with "No whitespace").

File: src/librustc_mir/transform/qualify_consts.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/transform/qualify_consts.rs]: Not sure about the "should" here, I was just very surprised to learn that this is how promotion works.

The discussion was at https://github.com/rust-lang/rust/pull/63955#discussion_r318491831.

**Comment by RalfJung** [comment]
> Would be nice if the PR author could toggle the default "whitespace" diff setting (the diff is way more readable with "No whitespace").

AFAIK this is a setting you pick, not something the PR author sets? I can toggle this behind the gear icon in the diff view.

**Comment by Centril** [comment]
@RalfJung Yeah I believe @eddyb was making a "feature request" to GitHub...

**Comment by RalfJung** [comment]
Oh I see. I read it as asking the PR author here to do something.^^

**Comment by Ixrec** [comment]
Not a great workaround, but: In addition to the setting, you can add `?w=1` to the `/files` url. So whenever I have a PR where hiding whitespace helps a lot, or only some commits should be reviewed because the others are autogenerated stuff or whatever, I usually put a "Recommend reviewing [without whitespace](https://github.com/rust-lang/rust/pull/63994/files?w=1)" link at the top of my PR description. I think I got the idea from Rust RFC PRs always having a "Rendered" link.

**Pull request merged**